services/blockstorage/adapter/leveldb_persistence.go

```diff
@@ -12,28 +30,138 @@ type levelDbBlockPersistence struct {
 	blockWritten chan bool
 	blockPairs   []*protocol.BlockPairContainer
 	config       Config
+	reporting    instrumentation.BasicLogger
+	db           *leveldb.DB
+}
+
+type config struct {
+	name string
+}
+
+func (c *config) NodeId() string {
+	return c.name
+}
+
+func NewLevelDbBlockPersistenceConfig(name string) Config {
+	return &config{name: name}
 }
 
 func NewLevelDbBlockPersistence(config Config) BlockPersistence {
+	db, err := leveldb.OpenFile("/tmp/db", nil)
+
+	if err != nil {
+		instrumentation.GetLogger(instrumentation.String("component", "persistence")).Error("Could not instantiate leveldb", instrumentation.Error(err))
+		panic("Could not instantiate leveldb")
+	}
+
 	return &levelDbBlockPersistence{
 		config:       config,
 		blockWritten: make(chan bool, 10),
+		db:           db,
+		reporting:    instrumentation.GetLogger(),
 	}
 }
 
+func (bp *levelDbBlockPersistence) WithLogger(reporting instrumentation.BasicLogger) BlockPersistence {
+	bp.reporting = reporting
+	return bp
+}
+
 func (bp *levelDbBlockPersistence) WriteBlock(blockPair *protocol.BlockPairContainer) {
-	bp.blockPairs = append(bp.blockPairs, blockPair)
+	var errors []error
+	var keys []string
+
+	if !basicValidation(blockPair) {
+		bp.reporting.Info("Block is invalid", instrumentation.Stringable("txBlockHeader", blockPair.TransactionsBlock.Header))
+		return
+	}
+
+	lastBlockHeight, lastBlockHeightRetrivalError := bp.loadLastBlockHeight()
+	errors = append(errors, lastBlockHeightRetrivalError)
+
+	txErrors, txKeys := bp.putTxBlock(blockPair.TransactionsBlock)
+	rsErrors, rsKeys := bp.putResultsBlock(blockPair.ResultsBlock)
+
+	newBlockHeight := blockPair.TransactionsBlock.Header.BlockHeight()
+	blockHeightSaveError := bp.saveLastBlockHeight(newBlockHeight)
+
+	errors = append(errors, blockHeightSaveError)
+
+	errors = append(errors, txErrors...)
+	errors = append(errors, rsErrors...)
+
+	keys = append(keys, txKeys...)
+	keys = append(keys, rsKeys...)
+
+	if hasErrors, firstError := anyErrors(errors...); hasErrors {
+		bp.saveLastBlockHeight(lastBlockHeight)
+		bp.reporting.Error("Failed to write block", instrumentation.BlockHeight(newBlockHeight), instrumentation.Error(firstError))
+
+		for _, key := range keys {
+			bp.revert(key)
+		}
+
+		bp.blockWritten <- false
+		return
+	}
+
 	bp.blockWritten <- true
 }
 
 func (bp *levelDbBlockPersistence) ReadAllBlocks() []*protocol.BlockPairContainer {
```

Comment:
This interface function can't stay in a persistent production implementation of real storage.. it's only applicable for memory games and short lived tests

We only wrote it as a hack to get the acceptance test working, we need to stop with it

Assuming we have 10GB of blocks, there's no way we return all of them as an array